# Part C1: Linear Regression — Normal Equation vs Gradient Descent vs Ill-Conditioning

Roll No: `CS26S510`

---

In [24]:
import numpy as np
import matplotlib.pyplot as plt

np.random.seed(0)
n = 300
x1 = np.random.uniform(-3, 3, n)
x2 = 0.98 * x1 + np.random.normal(0, 0.05, n)  # nearly collinear with x1
X = np.column_stack([x1, x2])
A = np.column_stack([np.ones(n), X])  # design matrix with intercept
theta_true = np.array([4.0, 3.0, -1.5])
y = A @ theta_true + np.random.normal(0, 1.0, n)
print(A.shape, y.shape)
print(A[:5])


(300, 3) (300,)
[[ 1.          0.29288102  0.29484873]
 [ 1.          1.2911362   1.27692253]
 [ 1.          0.61658026  0.57438285]
 [ 1.          0.2692991   0.25201703]
 [ 1.         -0.4580712  -0.52011283]]


## (a) Normal Equation from Scratch

Implement θ̂ = (AᵀA)⁻¹Aᵀy using numpy. Report θ̂ and training RMSE.

In [22]:
def normal_equation(A, y):
    """Closed-form OLS solution: theta_hat = (A^T A)^{-1} A^T y"""
    AtA = A.T @ A
    Aty = A.T @ y
    theta_hat = np.linalg.solve(AtA, Aty)
    return theta_hat

def rmse(A, theta, y):
    residuals = y - A @ theta
    return np.sqrt(np.mean(residuals**2))

theta_hat = normal_equation(A, y)
train_rmse = rmse(A, theta_hat, y)

print("theta_hat:", theta_hat)
print("true theta:", theta_true)
print("training RMSE:", train_rmse)


theta_hat: [3.88258777 1.08031127 0.48216857]
true theta: [ 4.   3.  -1.5]
training RMSE: 0.9683944858396789


**Discussion:** The estimated θ̂ = [3.88, 1.08, 0.48] deviates substantially from the true θ = [4.0, 3.0, -1.5] in the coefficients for x1 and x2, even though the training RMSE (≈0.97) is close to the true noise level (σ=1.0), indicating an excellent fit. This happens because x2 is constructed as a near-exact linear function of x1 (x2 ≈ 0.98·x1), making the two columns of A nearly linearly dependent. As a result, the least-squares solution cannot uniquely determine how much of the predictive effect belongs to x1 versus x2 — it only recovers their combined contribution accurately (θ̂₁·x1 + θ̂₂·x2 ≈ θ₁_true·x1 + θ₂_true·x2), while the individual coefficients are highly sensitive to noise and can differ greatly from the ground truth. The intercept θ̂₀, unaffected by this collinearity, is recovered accurately. This is a direct consequence of A being ill-conditioned, which we quantify next via cond(AᵀA).

## (b) Condition Number & Perturbation Sensitivity

Compute cond(AᵀA). Perturb y, recompute θ̂', report ‖ε_θ‖.

In [14]:
# TODO: implement this part


**Discussion:** _(write your 2-3 sentence interpretation here)_

## (c) Gradient Descent

Implement full-batch GD for 3 learning rates. Plot J(θ_t) vs iteration (log-scale y). Compare to (a).

In [15]:
# TODO: implement this part


**Discussion:** _(write your 2-3 sentence interpretation here)_

## (d) Ridge Regularized Normal Equation

Repeat (a)-(b) with θ̂_ridge = (AᵀA + λI)⁻¹Aᵀy, λ=1.0. Discuss sensitivity improvement.

In [16]:
# TODO: implement this part


**Discussion:** _(write your 2-3 sentence interpretation here)_